Do not create dagruns for DAGs with import errors

ephraimbuddy opened the pull request:
An active dag can suddenly have import errors as a result of the DAG file being changed. Currently, we
do not consider this before creating dagruns.

This PR adds the has_import_errors to `dag_model` so dags with import errors are not set to the scheduler to
create dagruns. Also, dag runs can no longer be created on the UI/API when the dag has import errors


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

kaxil reviewed (commented):
Can we add a test for this case please?

Separately let's check why `max_active_runs` is not set via tests too

jedcunningham commented on airflow/models/dag.py:
```diff
@@ -2743,7 +2743,9 @@ def __init__(self, concurrency=None, **kwargs):
-            self.max_active_runs = conf.getint('core', 'max_active_runs_per_dag')
+            self.max_active_runs = kwargs.get(
+                'max_active_runs', conf.getint('core', 'max_active_runs_per_dag', fallback=16)
+            )
```
Wait, why do we need the fallback? It should get it from the default config anyways.

ephraimbuddy replied:
I just added it to be safe in case max_active_runs_per_dag is removed from the configuration. 
Can't think of any other reason why they'd get that errror

jedcunningham replied:
Gotcha. They'd have to remove it from the `config_templates/default_airflow.cfg` in their install, so I think that is unlikely to be the cause.

kaxil commented on airflow/models/dag.py:
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
         # We limit so that _one_ scheduler doesn't try to do all the creation
         # of dag runs
+        # exclude dags with import error
+        import_errors = [c.filename for c in session.query(importerror.filename).all()]
```
Can you test this with zip file too please

ashb commented on airflow/models/dag.py:
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
         query = (
             session.query(cls)
             .filter(
+                cls.max_active_runs != None,  # noqa
```
What does this have to do with import errors?

ephraimbuddy replied:
Actually the cause of the issue was that an active dag had errors after upgrade, normally it should have max_active_runs but it didn’t have it because it failed to parse. 

I’ll update it to only include this check and remove import errors

ashb commented on airflow/models/dag.py:
```diff
@@ -2880,12 +2883,16 @@ def dags_needing_dagruns(cls, session: Session):
                 cls.is_paused == expression.false(),
                 cls.is_active == expression.true(),
                 cls.next_dagrun_create_after <= func.now(),
+                cls.fileloc.notin_(import_errors),
```
We should find another way of doing this -- this function is part of the "hot" path of the scheduler, and a) adding an extra query and b) sending (possibly) a large number of params via this "NOT IN (?,?,?)" etc is not a great idea.

ephraimbuddy replied:
I have updated it to exclude those with null max_active_runs.

ephraimbuddy committed:
Do not create dagruns for DAGs with import errors

An active dag can suddenly have import errors as a result of the DAG file being changed. Currently, we
do not consider this before creating dagruns.

This PR adds the has_import_errors to dagmodel so dags with import errors are not set to the scheduler to
create dagruns

fixup! Add has_import_errors to DagModel and ensure Dags with import errors are not sent to the Scheduler

fixup! fixup! Add has_import_errors to DagModel and ensure Dags with import errors are not sent to the Scheduler

Update airflow/models/dag.py

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

Update docs/apache-airflow/migrations-ref.rst

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

Apply suggestions from code review

Co-authored-by: Ash Berlin-Taylor <ash_github@firemirror.com>

fixup! Apply suggestions from code review

restrict manual dagrun creation if dagrun has import errors

fixup! restrict manual dagrun creation if dagrun has import errors

use session rollback

ephraimbuddy commented on tests/models/test_dag.py:
```diff
@@ -1848,9 +1888,6 @@ def test_dags_needing_dagruns_not_too_early(self):
         assert dag_models == []
 
-        session.rollback()
-        session.close()
```
Removed this because of the added clean_db method

ashb replied:
But with these rollbacks it _shouldn't_ need a clean method.

ephraimbuddy replied:
Yes. I was seeing that it was repeated, happy to return it back if it's preferred

jedcunningham commented on airflow/dag_processing/processor.py:
```diff
@@ -533,6 +533,10 @@ def update_import_errors(session: Session, dagbag: DagBag) -> None:
 
         # Add the errors of the processed files
         for filename, stacktrace in dagbag.import_errors.items():
+            dag_model = session.query(DagModel).filter(DagModel.fileloc == filename).first()
```
I don't think this can be on just the first DagModel, don't we need to set it on all of them from that file?

kaxil replied:
```suggestion
            session.query(DagModel).filter(DagModel.fileloc == filename).update({"has_import_errors": True})
```

might be better _I think_

ashb commented on airflow/models/dag.py:
```diff
@@ -2880,11 +2884,14 @@ def dags_needing_dagruns(cls, session: Session):
         query = (
             session.query(cls)
             .filter(
+                cls.max_active_runs != None,  # noqa
```
```suggestion
```

We don't need this condition now I think, right?

ephraimbuddy replied:
Yes. Will remove it

ashb commented on docs/apache-airflow/migrations-ref.rst:
```diff
@@ -23,7 +23,9 @@ Here's the list of all the Database Migrations that are executed via when you ru
 | Revision ID                    | Revises ID       | Airflow Version | Description                                                                           |
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
-| ``7b2661a43ba3`` (head)        | ``142555e44c17`` | ``2.2.0``       | Change ``TaskInstance`` and ``TaskReschedule`` tables from execution_date to run_id.  |
+| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.2``       | Add has_import_errors column to DagModel                                              |
```
```suggestion
| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.3``       | Add has_import_errors column to DagModel                                              |
```

ashb commented on tests/models/test_dag.py:
```diff
@@ -1863,13 +1900,29 @@ def test_max_active_runs_not_none(self):
+        session.add(orm_dag)
+        session.flush()
+
+        assert not orm_dag.has_import_errors
```
This test isn't really adding much -- it's testing `self.has_import_errors = False` in the constructor.

It certainly doesn't need a whole separate test case for this, it can be added as an assert somewhere else if we want it. You already test basically everything here in `test_dags_needing_dagruns_doesnot_send_dagmodel_with_import_errors`, so this whole test function isn't needed

ashb commented on tests/models/test_dag.py:
```diff
@@ -1898,8 +1951,65 @@ def test_dags_needing_dagruns_only_unpaused(self):
+
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == [orm_dag]
+        # set to None and ensure it's not sent to scheduler
```
```suggestion
```

ashb commented on tests/models/test_dag.py:
```diff
@@ -1898,8 +1951,65 @@ def test_dags_needing_dagruns_only_unpaused(self):
+        session.flush()
+        needed = DagModel.dags_needing_dagruns(session).all()
+        assert needed == []
+
```
```suggestion
```

I don't think we need this behaviour _and_ the explicit flag, so since we have added the flag we should just use that.

In which case we should also re-title this PR please.

ashb reviewed (dismissed):
Happy to discuss about if we still need the max_active_runs condition or not too.

ashb commented on docs/apache-airflow/migrations-ref.rst:
```diff
@@ -23,7 +23,9 @@ Here's the list of all the Database Migrations that are executed via when you ru
 | Revision ID                    | Revises ID       | Airflow Version | Description                                                                           |
 +--------------------------------+------------------+-----------------+---------------------------------------------------------------------------------------+
-| ``7b2661a43ba3`` (head)        | ``142555e44c17`` | ``2.2.0``       | Change ``TaskInstance`` and ``TaskReschedule`` tables from execution_date to run_id.  |
+| ``be2bfac3da23`` (head)        | ``7b2661a43ba3`` | ``2.2.3``       | Add has_import_errors column to DagModel                                              |
```
I'm tempted to say this should be 2.3.0 -- it's maybe a bug fix, but this behaviour has been the same for all of 2.x so far.

What do others think here?

potiuk replied:
I am for adding it to 2.2.3. There is hardly a use of DagRuns for DAGs with import errors, especially that you could not see whether there were errors via API. Even if it requires data model change, that's fine as patch-level. This is OK for a bugfix to require new field in the DB.

jedcunningham replied:
Since this issue will crash the scheduler, I lean toward `2.2.3`.

ashb reviewed (commented):
LGTM now, just the 2.2.3 vs 2.3 decision.

jedcunningham commented on tests/dag_processing/test_processor.py:
```diff
@@ -481,6 +481,34 @@ def test_add_unparseable_zip_file_creates_import_error(self, tmpdir):
+        assert len(import_errors) == 1
+        import_error = import_errors[0]
+        assert import_error.filename == temp_dagfile
+        assert import_error.stacktrace == f"invalid syntax ({TEMP_DAG_FILENAME}, line 53)"
```
```suggestion
```

nit: Do we need to assert the traceback?

ephraimbuddy replied:
I think there's no harm in asserting it?

jedcunningham replied:
At least in my view, it's needlessly fragile. since we append, that line number could change if that test dag is updated.

jedcunningham commented on tests/dag_processing/test_processor.py:
```diff
@@ -481,6 +481,34 @@ def test_add_unparseable_zip_file_creates_import_error(self, tmpdir):
+        import_error = import_errors[0]
+        assert import_error.filename == temp_dagfile
+        assert import_error.stacktrace == f"invalid syntax ({TEMP_DAG_FILENAME}, line 53)"
+        session.rollback()
```
I don't think this is actually doing anything - `update_import_errors` does a commit already. That said, this should probably stay and happen at the end of this test case.

jedcunningham commented on tests/models/test_dag.py:
```diff
@@ -822,6 +822,36 @@ def test_bulk_write_to_db_max_active_runs(self, state):
         model = session.query(DagModel).get((dag.dag_id,))
         assert model.next_dagrun_create_after is None
 
+    def test_bulk_write_to_db_has_import_error(self):
+        """
+        Test that DagModel.has_import_error is set to false if no import errors.
+        """
+        dag = DAG(dag_id='test_has_import_error', start_date=DEFAULT_DATE)
+
+        DummyOperator(task_id='dummy', dag=dag, owner='airflow')
+
+        session = settings.Session()
```
```suggestion
    def test_bulk_write_to_db_has_import_error(self, session):
        """
        Test that DagModel.has_import_error is set to false if no import errors.
        """
        dag = DAG(dag_id='test_has_import_error', start_date=DEFAULT_DATE)

        DummyOperator(task_id='dummy', dag=dag, owner='airflow')

```

Use the session fixture?

ephraimbuddy replied:
Ah, yeah. That's better

ephraimbuddy replied:
Turns out the class TestDag is using unittest.TestCase and not pytest. Wondering if it's worth it to convert it to pytest in this PR so we can use the session fixture or create a session fixture inside the unittest class as shown here?: https://pytest.org/en/latest/how-to/unittest.html#using-autouse-fixtures-and-accessing-other-fixtures

jedcunningham replied:
I'd say do that in another PR. Plenty of other tests could benefit from it as well.

jedcunningham commented on tests/models/test_dag.py:
```diff
@@ -1901,6 +1932,36 @@ def test_dags_needing_dagruns_only_unpaused(self):
+        assert needed == []
+
+        session.rollback()
+        session.close()
```
```suggestion
```

The session fixture already does a rollback.

ephraimbuddy committed:
Apply suggestions from code review

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

ephraimbuddy committed:
fixup! Apply suggestions from code review

ephraimbuddy committed:
fixup! Apply suggestions from code review
